pkg/auth/basic_auth.go

```diff
@@ -0,0 +1,74 @@
+package auth
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// UserPrincipal - represents a User as a Principal to the auth system.
+type UserPrincipal struct {
+	username string
+	// might need a set of permissions etc
+}
+
+// GetType - returns "user" indicating it is a UserPrincipal
+func (u UserPrincipal) GetType() string {
+	return "user"
+}
+
+// GetName - returns user's name
+func (u UserPrincipal) GetName() string {
+	return u.username
+}
+
+// BasicAuth - Performs an HTTP Basic Auth validation.
+type BasicAuth struct {
+	usa UserServiceAdapter
+}
+
+// NewBasicAuth - constructs a BasicAuth instance.
+func NewBasicAuth(userSvcAdapter UserServiceAdapter) BasicAuth {
+	return BasicAuth{usa: userSvcAdapter}
+}
+
+// GetPrincipal - returns the User Principal that matches the credentials in the
+// Authorization header.
+func (b BasicAuth) GetPrincipal(r *http.Request) (Principal, error) {
+	var username string
+	var password string
+
+	// get Authorization header
+	authheader := r.Header.Get("Authorization")
+	if strings.HasPrefix(strings.ToUpper(authheader), "BASIC ") {
+		// get the encoded part of the header
+		decodedheader, err := base64.StdEncoding.DecodeString(authheader[6:])
+		if err != nil {
+			fmt.Println(err.Error())
+		}
+		userpass := strings.Split(string(decodedheader), ":")
+		username = userpass[0]
```

Comment:
So I tried quite a few things in tests, all are `Authorization Basic ` with different encoded values:

1) just the username "admin"
2) just the password ":password"
3) empty string

all are handled correctly no panic on this line. `strings.Split` always returns a slice of at least one item even if that item is empty string.